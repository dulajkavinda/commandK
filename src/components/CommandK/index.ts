export { default as CommandK } from './CommandK'
