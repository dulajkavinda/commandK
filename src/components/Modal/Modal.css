@keyframes fadein {
  from {
    transform: scale(0.8);
    opacity: 0;
  }
  to {
    transform: scale(1);
    opacity: 1;
  }
}

@keyframes bg-color {
  from {
    background-color: transparent;
  }
  to {
    background-color: rgba(#1e262f, 0.6);
  }
}

.modal-overlay {
  z-index: 99999;
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  background: rgba(#1e262f, 0.7);
  display: flex;
  justify-content: center;
  align-items: center;
  animation: bg-color 0.1s;
}

.modal-box {
  border: 1px solid #494e54;

  padding-left: 0.5em;
  padding-right: 0.5em;
  font-family: sans-serif;
}

.modal-box-medium {
  display: block;
  background: #1e262f;
  width: 43em;
  min-height: 30%;
  border-radius: 1rem;
  animation: fadein 0.2s;
}

.modal-box-small {
  display: block;
  background: #1e262f;
  width: 30em;
  min-height: 30%;
  padding: 1rem;
  border-radius: 1rem;
  animation: fadein 0.2s;
}

.modal-box-large {
  display: block;
  background: #1e262f;
  width: 70%;
  min-height: 30%;
  padding: 1rem;
  border-radius: 1rem;
  animation: fadein 0.2s;
}

.modal-box-header {
  border-top-left-radius: 1rem;
  border-top-right-radius: 1rem;
  border-bottom: 1px solid #494e54;
  display: flex;
  height: 3em;
  width: 100%;
}

.modal-box-header-search {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
}

.modal-box-header-search-left {
  display: flex;
  flex-direction: row;
}

.modal-box-header-search-results {
  color: #8f8e8e;
  margin-right: 5px;
  font-size: 1rem;
}

.modal-box-header-search-icon {
  display: flex;
  align-items: center;
  margin-left: 5px;
  margin-top: 1px;
}

.modal-box-header-search-username {
  color: #8095a9;
  font-size: 0.9rem;
  font-weight: 500;
  font-family: sans-serif;
  margin-left: 10px;
  margin-right: 10px;
  font-weight: 600;
  padding: 0.2em 0.8em;
  border-radius: 3px;

  background-color: #363e47;
}

.modal-box-header-search-input {
  border: none;
  outline: none;
  background: none;
  color: #d3d3d3;
  font-size: 1rem;
}

.modal-box-body {
  display: flex;
  flex-direction: column;

  overflow: scroll;
  overflow-y: auto;
  overflow-x: hidden;
  height: 20em;
}

.modal-box-body-item {
  width: 100%;
}

.modal-box-body-item-title {
  color: #8095a9;
  font-size: 0.9rem;
  font-weight: 500;
  font-weight: 600;
  padding-top: 1em;
  padding-bottom: 0.2em;
  margin-bottom: 10px;
  margin-left: 10px;
}

.modal-box-body-items:focus {
  outline: none;
  border: none;
  border: 1px dashed #687585;
  display: flex;
  flex-direction: row;
  align-items: center;
  color: #d3d3d3;
  background-color: #18222c;
  opacity: 0.7;
  -webkit-transition-property: opacity;
  -webkit-transition-duration: 0.5s;
  -webkit-transition-timing-function: cubic-bezier(0.15, 0.44, 0.47, 0.99);
  transition-property: opacity;
  transition-duration: 0.5s;
  transition-timing-function: cubic-bezier(0.15, 0.44, 0.47, 0.99);
}

.modal-box-body-items {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  color: #d3d3d3;
  padding: 0.8em 1em;
  border-radius: 10px;
  border: 1px solid transparent;
  cursor: pointer;
  background-color: #19222c;
  margin-top: 10px;
  margin-bottom: 10px;
  margin-right: 10px;
}

.modal-box-body-items:hover {
  opacity: 0.7;
  border: 1px dashed #687585;

  -webkit-transition-property: opacity;
  -webkit-transition-duration: 0.5s;
  -webkit-transition-timing-function: cubic-bezier(0.15, 0.44, 0.47, 0.99);
  transition-property: opacity;
  transition-duration: 0.5s;
  transition-timing-function: cubic-bezier(0.15, 0.44, 0.47, 0.99);
}

.modal-box-body-items-title {
  font-size: 1.1rem;
  font-weight: 600;
}

.modal-box-body-items-icon {
  margin-right: 10px;
}

.modal-box-body-items-left {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.modal-box-body-items-goto {
  display: flex;
  align-items: center;
  font-size: 0.9em;
  font-weight: 600;
  color: #687585;
}

.modal-box-body-items-goto-icon {
  margin-right: 5px;
  display: flex;
  align-items: center;
}

.modal-box-body-items-goto-text-start {
  margin-right: 5px;
}

.modal-box-body-items:not(:focus) .modal-box-body-items-goto {
  display: none;
}

.modal-box-header-search-right {
  display: flex;
  flex-direction: row;
}

.modal-box-header-search-right-nav {
  display: flex;
  flex-direction: row;

  justify-content: space-between;
  margin-right: 20px;
}

.modal-box-header-search-right-icon {
  margin-right: 7px;
}

.modal-box-header-search-right-text {
  font-size: 1em;
  font-weight: 600;
  color: #687585;
}

.modal-box-header-search-right-text-start {
  font-size: 1em;
  font-weight: 600;
  color: #687585;
  margin-right: 7px;
}

.modal-box-body::-webkit-scrollbar-thumb {
  background-color: #5d646b;
  border-radius: 20px;
}

.modal-box-body::-webkit-scrollbar {
  width: 8px;
}

.modal-box-body::-webkit-scrollbar-track {
  background: #1e262f;
}

/* mobile media queries */

@media (max-width: 768px) {
  .modal-overlay {
    align-items: end;
  }

  .modal-box-medium {
    width: 100%;
  }

  .modal-box-small {
    width: 100%;
  }

  .modal-box-large {
    width: 100%;
  }

  .modal-box-header-search-right-nav {
    display: none;
  }

  .modal-box-body-items-goto {
    display: none;
  }
}
