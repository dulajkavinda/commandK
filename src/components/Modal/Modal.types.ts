import { ReactNode } from 'react'

export type ModalSize = 'small' | 'medium' | 'large'

export interface ModalType {
  children?: ReactNode
  size?: ModalSize
  username?: string
  isOpen: boolean
  data: Group[]
  toggle: () => void
}
export interface State {
  selectedIndex: number
}

export type Action = { type: 'arrowUp' } | { type: 'arrowDown' } | { type: 'select'; payload: number }

export interface Item {
  icon: string
  title: string
  url?: string
}

export interface Group {
  sectionName: string
  items: Item[]
}
