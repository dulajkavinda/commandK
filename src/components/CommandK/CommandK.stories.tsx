import type { Meta, StoryObj } from '@storybook/react'
import CommandK from './CommandK'
import { Group } from '../Modal/Modal.types'

const meta = {
  title: 'CommandK',
  component: CommandK,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {},
} satisfies Meta<typeof CommandK>

export default meta

type Story = StoryObj<typeof meta>

const listA: Group[] = [
  {
    sectionName: 'Repositories',
    items: [
      {
        icon: '🗂',
        title: 'How to make a good repo',
        url: '/repo',
      },
      {
        icon: '🚀',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
    ],
  },
  // {
  //   sectionName: 'Projects',
  //   items: [
  //     {
  //       icon: '🏋️‍♀️',
  //       title: 'How to make a good repo',
  //     },
  //     {
  //       icon: '📸',
  //       title: 'This is another one to check',
  //     },
  //   ],
  // },
  // {
  //   sectionName: 'Something',
  //   items: [
  //     {
  //       icon: '🏋️‍♀️',
  //       title: 'How to make a good repo',
  //       url: 'https://google.lk',
  //     },
  //     {
  //       icon: '📸',
  //       title: 'This is another one to check',
  //       url: 'https://google.lk',
  //     },
  //     {
  //       icon: '🏋️‍♀️',
  //       title: 'How to make a good repo',
  //       url: 'https://google.lk',
  //     },
  //     {
  //       icon: '📸',
  //       title: 'This is another one to check',
  //       url: 'https://google.lk',
  //     },
  //   ],
  // },
]

export const Small: Story = {
  args: {
    keyLetter: 'K',
    modalSize: 'small',
    buttonSize: 'small',
    data: listA,
  },
}

export const Medium: Story = {
  args: {
    keyLetter: 'K',
    modalSize: 'medium',
    buttonSize: 'medium',
    data: listA,
    customStyles: {
      opacity: 0.9,
    },
  },
}

export const Large: Story = {
  args: {
    keyLetter: 'K',
    modalSize: 'large',
    buttonSize: 'large',
    data: listA,
  },
}
