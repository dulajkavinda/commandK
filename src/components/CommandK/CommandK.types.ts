import { ButtonSize } from '../Button/Button.types'
import { Group, ModalSize } from '../Modal/Modal.types'

export interface CommandKProps {
  keyLetter?: string
  buttonSize?: ButtonSize
  modalSize?: ModalSize
  data: Group[]
  username?: string
  perSectionLimit?: number
  hideButton?: boolean
  customStyles?: React.CSSProperties
}
