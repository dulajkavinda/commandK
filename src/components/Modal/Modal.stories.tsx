import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import Modal from './Modal'
import { DownArrowIcon } from './utils'
import { Group } from './Modal.types'

const meta = {
  title: 'Modal',
  component: Modal,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {},
} satisfies Meta<typeof Modal>

export default meta

type Story = StoryObj<typeof meta>

const listA: Group[] = [
  {
    sectionName: 'Repositories',
    items: [
      {
        icon: <DownArrowIcon size={14} />,
        title: 'How to make a good repo',
        url: '/repo',
      },
      {
        icon: '🚀',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
      {
        icon: '🗂',
        title: 'How to make a good repo',
        url: '/repo',
      },
      {
        icon: '🚀',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
    ],
  },
  {
    sectionName: 'Projects',
    items: [
      {
        icon: '🏋️‍♀️',
        title: 'How to make a good repo',
      },
      {
        icon: '📸',
        title: 'This is another one to check',
      },
    ],
  },
  {
    sectionName: 'Something',
    items: [
      {
        icon: '🏋️‍♀️',
        title: 'How to make a good repo',
        url: 'https://google.lk',
      },
      {
        icon: '📸',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
      {
        icon: '🏋️‍♀️',
        title: 'How to make a good repo',
        url: 'https://google.lk',
      },
      {
        icon: '📸',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
    ],
  },
]

export const Small: Story = {
  args: {
    data: listA,
    isOpen: true,
    toggle: () => {},
    username: 'home',
    size: 'small',
  },
}

export const Medium: Story = {
  args: {
    data: listA,
    isOpen: true,
    toggle: () => {},
    username: 'home',
    perSectionLimit: 3,
    size: 'medium',
  },
}

export const Large: Story = {
  args: {
    data: listA,
    isOpen: true,
    toggle: () => {},
    username: 'home',
    size: 'large',
  },
}
