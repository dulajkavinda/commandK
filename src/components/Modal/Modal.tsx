import React, { useEffect, useReducer, useRef, useMemo, useState } from 'react'
import classNames from 'classnames'
import { SearchIcon } from './SearchIcon'
import useKeyPress from '../../hooks/useKeyPress'
import { Action, Group, Item, ModalSize, ModalType, State } from './Modal.types'
import './Modal.css'

const Modal = (props: ModalType) => {
  const modalClass = classNames('modal-box', {
    [`modal-box-${props.size}`]: props.size,
  })

  const [list, setList] = useState<Group[] | []>(props.data || [])

  const inputRef = useRef<HTMLInputElement>(null)

  const arrowUpPressed = useKeyPress('ArrowUp')
  const arrowDownPressed = useKeyPress('ArrowDown')

  const initialState = { selectedIndex: 0 }

  const items = useMemo(() => {
    return list && list.map((item) => item.items).flat()
  }, [list])

  const filterItems = (list: Group[], title: string): void => {
    const result: Group[] = []

    list.forEach((obj) => {
      const items = obj.items.filter((item) => item.title.toLowerCase().includes(title.toLowerCase()))

      if (items.length > 0) {
        result.push({
          ...obj,
          items,
        })
      }
    })

    setList(result)
  }

  const reducer = (state: State, action: Action): State => {
    switch (action.type) {
      case 'arrowUp': {
        return {
          selectedIndex: state.selectedIndex,
        }
      }
      case 'arrowDown': {
        const button = document.getElementsByClassName(`modal-box-body-items-${state.selectedIndex}`)[0] as HTMLElement

        if (button) {
          button.focus()
        }

        return {
          selectedIndex: state.selectedIndex !== items.length - 1 ? state.selectedIndex + 1 : 0,
        }
      }
      case 'select':
        return { selectedIndex: action.payload }
      default:
        throw new Error()
    }
  }

  const [, dispatch] = useReducer(reducer, initialState)

  useEffect(() => {
    if (inputRef.current) {
      const x = window.scrollX
      const y = window.scrollY
      inputRef.current.focus({
        preventScroll: true,
      })
      window.scrollTo(x, y)
    }
  }, [])

  useEffect(() => {
    if (arrowUpPressed) {
      dispatch({ type: 'arrowUp' })
    }
  }, [arrowUpPressed])

  useEffect(() => {
    if (arrowDownPressed) {
      dispatch({ type: 'arrowDown' })
    }
  }, [arrowDownPressed])

  let trackItemindex = -1

  return (
    <>
      {props.isOpen && (
        <div className='modal-overlay' onClick={props.toggle}>
          <div onClick={(e) => e.stopPropagation()} id='modal-box' className={modalClass}>
            <div className='modal-box-header'>
              <div className='modal-box-header-search'>
                <div className='modal-box-header-search-left'>
                  <div className='modal-box-header-search-icon'>
                    <SearchIcon />
                    <span className='modal-box-header-search-username'>{props.username}</span>
                  </div>
                  <input
                    onChange={(e) => {
                      filterItems(props.data, e.target.value)
                    }}
                    ref={inputRef}
                    className='modal-box-header-search-input'
                    type='text'
                    placeholder='Search or jump to'
                    onKeyDown={(e) => {
                      if (e.key === 'ArrowDown') {
                        console.log('arrow down')
                      }
                    }}
                  />
                </div>

                <div className='modal-box-header-search-results'>{items.length} Results</div>
              </div>
            </div>
            <div className='modal-box-body'>
              {list &&
                list.map((item: Group, index) => {
                  return (
                    <div key={index} className='modal-box-body-item'>
                      <div className='modal-box-body-item-title'>{item.sectionName}</div>
                      {item.items.map((subItem: Item) => {
                        const listItemIndex = trackItemindex++
                        return (
                          <div
                            role='button'
                            tabIndex={0}
                            key={trackItemindex}
                            onKeyDown={(e) => {
                              if (e.key === 'Enter') {
                                window.location.href = `${subItem.url}`
                              }
                            }}
                            onClick={() => {
                              window.location.href = `${subItem.url}`
                            }}
                            className={classNames('modal-box-body-items', {
                              [`modal-box-body-items-${listItemIndex + 1}`]: true,
                            })}
                          >
                            <div className='modal-box-body-items-left'>
                              <div className='modal-box-body-items-icon'>{subItem.icon}</div>
                              <div className='modal-box-body-items-title'>{subItem.title}</div>
                            </div>
                            <span className='modal-box-body-items-goto'>Go to</span>
                          </div>
                        )
                      })}
                    </div>
                  )
                })}
            </div>
          </div>
        </div>
      )}
    </>
  )
}

Modal.defaultProps = {
  size: 'medium' as ModalSize,
  data: [],
}

export default Modal
