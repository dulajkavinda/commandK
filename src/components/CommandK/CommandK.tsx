import React, { useState, useEffect } from 'react'
import Modal from '../Modal/Modal'
import Button from '../Button/Button'
import { CommandKProps } from './CommandK.types'
import { ButtonSize } from '../Button/Button.types'

const CommandK = (props: CommandKProps) => {
  const [isOpen, setIsOpen] = useState(false)

  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      if ((event.metaKey && event.key === 'k') || (event.ctrlKey && event.key === 'k')) {
        const button = document.getElementById('search-button') as HTMLButtonElement | null
        if (button) {
          const x = window.scrollX
          const y = window.scrollY
          setIsOpen(!isOpen)
          button.focus({
            preventScroll: true,
          })
          window.scrollTo(x, y)
        }
      }
    }

    document.addEventListener('keydown', handleKeyDown)

    // prevents scrolling when modal is open
    if (isOpen && typeof window != 'undefined' && window.document) {
      document.body.style.overflow = 'hidden'
    }

    if (!isOpen) {
      document.body.style.overflow = ''
      return () => {
        document.removeEventListener('keydown', handleKeyDown)
      }
    }
  }, [isOpen])

  return (
    <div
      style={{
        height: '2000px',
      }}
      data-testid='commandk'
    >
      {isOpen && (
        <Modal
          data={props.data}
          size={props.modalSize}
          isOpen={isOpen}
          username={props.username}
          toggle={() => setIsOpen(false)}
        />
      )}
      <Button
        styles={props.customStyles}
        size={props.buttonSize}
        keyLetter={props.keyLetter}
        onClick={() => setIsOpen(true)}
      />
    </div>
  )
}

CommandK.defaultProps = {
  keyLetter: 'K',
  buttonSize: 'medium' as ButtonSize,
  username: 'home',
  customStyles: {},
}

export default CommandK
