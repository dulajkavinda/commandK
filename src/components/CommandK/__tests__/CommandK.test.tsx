import React from 'react'
import { screen, render, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import CommandK from '../CommandK'
import { Group } from '../../Modal/Modal.types'

const list: Group[] = [
  {
    sectionName: 'Repositories',
    items: [
      {
        icon: '🗂',
        title: 'How to make a good repo',
        url: '/repo',
      },
      {
        icon: '🚀',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
    ],
  },
  {
    sectionName: 'Projects',
    items: [
      {
        icon: '🏋️‍♀️',
        title: 'How to make a good repo',
      },
      {
        icon: '📸',
        title: 'This is another one to check',
      },
    ],
  },
  {
    sectionName: 'Something',
    items: [
      {
        icon: '🏋️‍♀️',
        title: 'How to make a good repo',
        url: 'https://google.lk',
      },
      {
        icon: '📸',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
      {
        icon: '🏋️‍♀️',
        title: 'How to make a good repo',
        url: 'https://google.lk',
      },
      {
        icon: '📸',
        title: 'This is another one to check',
        url: 'https://google.lk',
      },
    ],
  },
]

describe('components/CommandK', () => {
  it('renders commandK', () => {
    const modal = renderer.create(<CommandK data={list} />).toJSON()
    expect(modal).toMatchSnapshot()
  })

  it('renders with a button', () => {
    render(<CommandK data={list} />)
    const btn = screen.getByTestId('search-button')

    expect(btn).toHaveClass('search-bar')
    expect(btn).toHaveTextContent('K')
  })

  it('renders with a given key', () => {
    render(<CommandK data={list} keyLetter='N' />)
    const btn = screen.getByTestId('search-button')

    expect(btn).toHaveClass('search-bar')
    expect(btn).toHaveTextContent('N')
  })

  it('renders with a given size', () => {
    render(<CommandK data={list} buttonSize='small' modalSize='small' />)
    const btn = screen.getByTestId('search-button')

    expect(btn).toHaveClass('search-bar-small')
  })

  it('open modal when the button is pressed', () => {
    render(<CommandK data={list} />)
    const btn = screen.getByTestId('search-button')

    fireEvent.click(btn)

    const modal = screen.getByTestId('modal')
    expect(modal).toBeInTheDocument()
  })

  it('modal search input should be focused when K and Command is pressed', () => {
    render(<CommandK data={list} />)

    const container = screen.getByTestId('commandk')

    fireEvent.keyDown(container, {
      key: 'k',
      code: 'KeyK',
      ctrlKey: true,
    })

    const btn = screen.getByTestId('search-input')
    expect(btn).toHaveFocus()
  })
})
