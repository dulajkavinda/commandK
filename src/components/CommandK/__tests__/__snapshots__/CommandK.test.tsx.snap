// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CommandK renders commandK 1`] = `
<div
  data-testid="commandk"
>
  <button
    className="search-bar search-bar-medium"
    data-testid="search-button"
    id="search-button"
    onClick={[Function]}
    style={{}}
    type="button"
  >
    <div
      className="search-bar-icon search-bar-icon-medium"
    >
      <span>
        Ctrl +
      </span>
    </div>
    <div
      className="search-bar-key search-bar-key-medium"
    >
      K
    </div>
  </button>
</div>
`;
